File: SICP/Lecture_6_full/Code/Java/Stream/src/fayelab/sicp/stream/nostream/BtQueens.java
package fayelab.sicp.stream.nostream;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static java.util.Arrays.*;

public class BtQueens
{
    public static Optional<List<List<Integer>>> solveOne(int n)
    {
        List<List<Integer>> solution = emptyBoard();
        solveOne(n, 1, 1, solution);
        return solution.isEmpty() ? Optional.empty() : Optional.of(solution);
    }

    private static void solveOne(int size, int curRow, int curCol, List<List<Integer>> curPoses)
    {
        if(curRow > size)
        {
            return;
        }
        
        if(curCol > size)
        {
            if(curPoses.isEmpty())
            {
                return;
            }
            
            List<Integer> lastPos = curPoses.remove(curPoses.size() - 1);
            solveOne(size, lastPos.get(0), lastPos.get(1) + 1, curPoses);
            return;
        }
        
        adjoinPos(curRow, curCol, curPoses);
        if(isSafe(curPoses))
        {
            solveOne(size, curRow + 1, 1, curPoses);
            return;
        }
        
        curPoses.remove(curPoses.size() - 1);
        solveOne(size, curRow, curCol + 1, curPoses);
    }
    
    public static Optional<List<List<List<Integer>>>> solveAll(int n)
    {
        List<List<List<Integer>>> solutions = new ArrayList<>();
        solveAll(n, 1, 1, emptyBoard(), solutions);
        return solutions.isEmpty() ? Optional.empty() : Optional.of(solutions);
    }

    private static void solveAll(int size, int curRow, int curCol, 
            List<List<Integer>> curPoses, List<List<List<Integer>>> solutions)
    {
        if(curRow > size)
        {
            solutions.add(new ArrayList<>(curPoses));
            List<Integer> lastPos = curPoses.remove(curPoses.size() - 1);
            solveAll(size, lastPos.get(0), lastPos.get(1) + 1, curPoses, solutions);
            return;
        }
        
        if(curCol > size)
        {
            if(curPoses.isEmpty())
            {
                return;
            }
            
            List<Integer> lastPos = curPoses.remove(curPoses.size() - 1);
            solveAll(size, lastPos.get(0), lastPos.get(1) + 1, curPoses, solutions);
            return;
        }
        
        adjoinPos(curRow, curCol, curPoses);
        if(isSafe(curPoses))
        {
            solveAll(size, curRow + 1, 1, curPoses, solutions);
            return;
        }
        
        curPoses.remove(curPoses.size() - 1);
        solveAll(size, curRow, curCol + 1, curPoses, solutions);
    }
    
    private static List<List<Integer>> emptyBoard()
    {
        return new ArrayList<>();
    }

    private static void adjoinPos(int curRow, int curCol, List<List<Integer>> curPoses)
    {
        curPoses.add(asList(curRow, curCol));
    }
    
    private static boolean isSafe(List<List<Integer>> poses)
    {
        List<Integer> newPos = poses.get(poses.size() - 1);
        List<List<Integer>> restPoses = poses.subList(0, poses.size() - 1);
        return isNotSameCol(newPos, restPoses) && isNotDiagonal(newPos, restPoses);
    }

    private static boolean isNotSameCol(List<Integer> newPos, List<List<Integer>> restPoses)
    {
        return restPoses.stream()
                        .noneMatch(restPos -> newPos.get(1) == restPos.get(1));
    }

    private static boolean isNotDiagonal(List<Integer> newPos, List<List<Integer>> restPoses)
    {
        return restPoses.stream()
                        .noneMatch(restPos -> Math.abs(newPos.get(0) - restPos.get(0)) 
                                                == Math.abs(newPos.get(1) - restPos.get(1)));
    }
}

File: SICP/Lecture_6_full/Code/Python/nostream/btqueens.py
## Eight Queens Problem, Backtracking Search
import copy

def solveone(n):
    solution = emptyboard()
    _solveone(n, 1, 1, solution)
    return solution if solution != [] else None

def _solveone(size, currow, curcol, curposes):
    if currow > size:
        return

    if curcol > size:
        if curposes == []:
            return

        lastpos = curposes.pop()
        _solveone(size, lastpos[0], lastpos[1] + 1, curposes)
        return

    adjoinpos(currow, curcol, curposes)
    if issafe(curposes):
        _solveone(size, currow + 1, 1, curposes)
        return

    curposes.pop()
    _solveone(size, currow, curcol + 1, curposes)

def solveall(n):
    solutions = []
    _solveall(n, 1, 1, emptyboard(), solutions)
    return solutions if solutions != [] else None

def _solveall(size, currow, curcol, curposes, solutions):
    if currow > size:
        solutions.append(copy.copy(curposes))
        lastpos = curposes.pop()
        _solveall(size, lastpos[0], lastpos[1] + 1, curposes, solutions)
        return

    if curcol > size:
        if curposes == []:
            return

        lastpos = curposes.pop()
        _solveall(size, lastpos[0], lastpos[1] + 1, curposes, solutions)
        return

    adjoinpos(currow, curcol, curposes)
    if issafe(curposes):
        _solveall(size, currow + 1, 1, curposes, solutions)
        return

    curposes.pop()
    _solveall(size, currow, curcol + 1, curposes, solutions)

def emptyboard():
    return [];

def adjoinpos(currow, curcol, curposes):
    curposes.append((currow, curcol))

def issafe(poses):
    newpos = poses[-1]
    restposes = poses[:-1]
    return is_not_samecol(newpos, restposes) and is_not_diagonal(newpos, restposes)

def is_not_samecol(newpos, restposes):
    return all(map(lambda restpos: newpos[1] != restpos[1], restposes))

def is_not_diagonal(newpos, restposes):
    return all(map(lambda restpos: abs(newpos[0] - restpos[0]) != abs(newpos[1] - restpos[1]), restposes))

## Tests
def test_solveone():
    assert [(1, 2), (2, 4), (3, 1), (4, 3)] == solveone(4)
    assert None == solveone(3)
    assert [(1, 1), (2, 5), (3, 8), (4, 6), 
            (5, 3), (6, 7), (7, 2), (8, 4)] == solveone(8)

    print 'test_solveone ok'

def test_solveall():
    assert [[(1, 2), (2, 4), (3, 1), (4, 3)],
            [(1, 3), (2, 1), (3, 4), (4, 2)]] == solveall(4)
    assert None == solveall(3)
    # assert 92 == len(solveall(8))

    print 'test_solveall ok'

def test():
    test_solveone()
    test_solveall()

    print 'test ok'

if __name__ == '__main__':
    test()